R/pkg/R/DataFrame.R

```diff
@@ -2199,3 +2199,97 @@ setMethod("coltypes",
 
             rTypes
           })
+
+#' Display the structure of a DataFrame, including column names, column types, as well as a
+#' a small sample of rows.
+#' @name str
+#' @title Compactly display the structure of a dataset
+#' @rdname str
+#' @family DataFrame functions
+#' @param object a DataFrame
+#' @examples \dontrun{
+#' # Create a DataFrame from the Iris dataset
+#' irisDF <- createDataFrame(sqlContext, iris)
+#' 
+#' # Show the structure of the DataFrame
+#' str(irisDF)
+#' }
+setMethod("str",
+          signature(object = "DataFrame"),
+          function(object) {
+
+            # TODO: These could be made global parameters, though in R it's not the case
+            MAX_CHAR_PER_ROW <- 120
+            MAX_COLS <- 100
+
+            # Get the column names and types of the DataFrame
+            names <- names(object)
+            types <- coltypes(object)
+
+            # Get the number of rows.
+            # TODO: Ideally, this should be cached
+            cachedCount <- nrow(object)
+
+            # Get the first elements of the dataset. Limit number of columns accordingly
+            dataFrame <- if (ncol(object) > MAX_COLS) {
+                           head(object[, c(1:MAX_COLS)])
+                         } else {
+                           head(object)
+                         }
+
+            # The number of observations will be displayed only if the number
+            # of rows of the dataset has already been cached.
+            if (!is.null(cachedCount)) {
```

Comment:
Yes, that's why I added the TODO. In our implementation, we had a global cache to store the number of rows of the datasets in the current session. At some point, we'll need to implement some caching mechanism so that every time you run str() or nrow(), you don't have to do a full data scan. When such caching mechanism is implemented, all we'll need to do is to change this line accordingly:

cachedCount <- nrow(object)

by

cachedCount <- FUNCTION_TO_GET_CACHED_NROW(object)

The behavior of str() is such that if nrow() hasn't been cached, the number of rows is simply not shown.